### ISSUE TITLE
fast训练无法生成预览图

### ISSUE THREAD

**Issue by soutsukimaru** [wochenlong/lora-scripts-next#126]
<img width="1290" height="289" alt="Image" src="https://github.com/user-attachments/assets/ccebcfe7-d339-4164-9ac9-42e6bf220659" />

<img width="1590" height="502" alt="Image" src="https://github.com/user-attachments/assets/9df695ab-3e04-44a0-a1e8-535fadb8fd6b" />
更新前正常，更新后训练中无法生成预览图

**Comment by MikumikuDAIFans** [comment]
代码修改要点：
- Anima Fast adapter 对缓存参数做了明确归一：禁用 live cache 时同步关闭 `cache_*_to_disk`，避免 Fast 运行时读取不完整 cache。
- 对 `skip_cache_check` 与 cache 参数冲突增加自动降级，改为 live encoding 路径。
- 为 Fast 训练设置更短、更稳定的日志默认项 `log_prefix=af_`、`log_tracker_name=tb`。
- 在写入 Fast TOML 前预创建日志 tracker 目录，规避 TensorBoard/长路径目录不存在导致的训练或预览问题。
- 增加 Fast adapter/cache/log 相关单元测试。

当前判断：在本地 Fast 短训和预览验证范围内有效；不关闭 issue，等待代码推送和团队复核。

**Comment by wochenlong** [comment]
已提交 PR 处理其中一个高概率根因：\max_train_epochs=1\ 且 \sample_every_n_epochs=2\ 时，训练结束前不会触发采样，导致 Fast 预览图为空。

**PR**：待创建后补充链接

**改动（保守）**：
- 预览开启时，将 \sample_every_n_epochs\ clamp 到不超过 \max_train_epochs\
- 未填 interval 时默认 \min(2, max_train_epochs)\
- 不自动改 \sample_at_first\；若使用 \sample_every_n_steps\ 则不干预

Issue 暂不关闭：仍需确认用户是否已打开「启用训练预览图」、训练监控是否点了「开启预览」，以及其它环境相关因素。合并后欢迎继续反馈。

**Comment by wochenlong** [comment]
PR 链接：https://github.com/wochenlong/lora-scripts-next/pull/156

**Comment by soutsukimaru** [comment]
<img width="1204" height="81" alt="Image" src="https://github.com/user-attachments/assets/7fbb0ef1-5630-4995-acea-9957f5ec6ee5" />

<img width="1200" height="272" alt="Image" src="https://github.com/user-attachments/assets/2c4a5c25-7717-46e2-9b95-3ed61c595bcf" />

<img width="1221" height="79" alt="Image" src="https://github.com/user-attachments/assets/688906eb-9c6a-48af-9ea0-05a00088f2e4" />

<img width="1244" height="89" alt="Image" src="https://github.com/user-attachments/assets/e08cbe53-39f6-4d23-a30a-071c6ba00984" />

<img width="1040" height="107" alt="Image" src="https://github.com/user-attachments/assets/51a86995-d7d5-4f79-aedb-40f540064620" />
如图所示，\max_train_epochs=30\， \sample_every_n_epochs=1\，当epoch训练至3时依旧未生成预览图

<img width="1127" height="640" alt="Image" src="https://github.com/user-attachments/assets/785f1374-19e5-4197-abe1-78db42f6a9a2" />
已打开「启用训练预览图」、训练监控已「开启预览」
Next Trainer版本为v2.8.0

**Comment by soutsukimaru** [comment]
以下为补充的训练日志
                             count: 1                                           
                    INFO     bucket 2: resolution (1248, 832),       base.py:545
                             count: 5                                           
                    INFO     mean ar error (without repeats): 0.0    base.py:553
                    INFO     caption dropout DISABLED (rate=0.0 on train.py:1505
                             all subsets)                                       
                    INFO     preparing accelerator                 train.py:2072
accelerator device: cuda
2026-06-24 03:35:51 INFO     Loading Qwen3 text encoder...          train.py:460
                    INFO     Loading Qwen3 text encoder from      weights.py:386
                             E:/AI/ComfyUI-aki-v3/ComfyUI/models/               
                             text_encoders/qwen_3_06b_base.safete               
                             nsors                                              
2026-06-24 03:35:58 INFO     Loaded Qwen3 state dict: <All keys   weights.py:439
                             matched successfully>                              
                    INFO     Loaded Qwen3 text encoder.           weights.py:447
                             Parameters: 596,049,920                            
                    INFO     Loading Anima VAE...                   train.py:474
                    INFO     Initializing VAE                   qwen_vae.py:1870
                    INFO     Loading VAE from                   qwen_vae.py:1894
                             E:/AI/ComfyUI-aki-v3/ComfyUI/model                 
                             s/vae/qwen_image_vae.safetensors                   
                    INFO     Converted ComfyUI AutoencoderKL    qwen_vae.py:1799
                             state dict keys to official format                 
                    INFO     Loaded VAE: <All keys matched      qwen_vae.py:1901
                             successfully>                                      
2026-06-24 03:35:59 INFO     Using attention mode: torch            train.py:522
                    INFO     Loading Anima DiT model with           train.py:542
                             attn_softmax_scale: None...                        
                    INFO     Loading DiT model from               weights.py:160
                             E:/AI/ComfyUI-aki-v3/ComfyUI/models/               
                             diffusion_models/anima_baseV10.safet               
                             ensors, device=cuda                                
                    INFO     Loading model files:              lora_utils.py:163
                             ['E:/AI/ComfyUI-aki-v3/ComfyUI/mo                  
                             dels/diffusion_models/anima_baseV                  
                             10.safetensors']                                   
                    INFO     Loading state dict. Dtype of      lora_utils.py:320
                             weight: torch.bfloat16, hook                       
                             enabled: False                                     

Loading anima_baseV10.safetensors:   0%|          | 0/545 [00:00<?, ?it/s]
Loading anima_baseV10.safetensors:   2%|▏         | 11/545 [00:00<00:05, 96.09it/s]
Loading anima_baseV10.safetensors:   5%|▍         | 26/545 [00:00<00:04, 113.74it/s]
Loading anima_baseV10.safetensors:   8%|▊         | 41/545 [00:00<00:04, 122.08it/s]
Loading anima_baseV10.safetensors:  10%|█         | 56/545 [00:00<00:03, 122.55it/s]
Loading anima_baseV10.safetensors:  13%|█▎        | 73/545 [00:00<00:03, 134.30it/s]
Loading anima_baseV10.safetensors:  16%|█▌        | 88/545 [00:00<00:03, 134.27it/s]
Loading anima_baseV10.safetensors:  19%|█▉        | 103/545 [00:00<00:03, 134.02it/s]
Loading anima_baseV10.safetensors:  22%|██▏       | 118/545 [00:00<00:03, 134.33it/s]
Loading anima_baseV10.safetensors:  26%|██▌       | 141/545 [00:01<00:02, 161.22it/s]
Loading anima_baseV10.safetensors:  29%|██▉       | 160/545 [00:01<00:02, 165.20it/s]
Loading anima_baseV10.safetensors:  33%|███▎      | 178/545 [00:01<00:02, 166.18it/s]
Loading anima_baseV10.safetensors:  36%|███▌      | 195/545 [00:01<00:02, 159.44it/s]
Loading anima_baseV10.safetensors:  39%|███▉      | 212/545 [00:01<00:02, 157.61it/s]
Loading anima_baseV10.safetensors:  42%|████▏     | 228/545 [00:01<00:02, 149.08it/s]
Loading anima_baseV10.safetensors:  46%|████▌     | 249/545 [00:01<00:01, 164.81it/s]
Loading anima_baseV10.safetensors:  49%|████▉     | 266/545 [00:01<00:01, 160.20it/s]
Loading anima_baseV10.safetensors:  53%|█████▎    | 288/545 [00:01<00:01, 175.76it/s]
Loading anima_baseV10.safetensors:  56%|█████▋    | 307/545 [00:02<00:01, 178.32it/s]
Loading anima_baseV10.safetensors:  60%|█████▉    | 325/545 [00:02<00:01, 139.46it/s]
Loading anima_baseV10.safetensors:  63%|██████▎   | 341/545 [00:02<00:01, 132.34it/s]
Loading anima_baseV10.safetensors:  65%|██████▌   | 356/545 [00:02<00:01, 128.91it/s]
Loading anima_baseV10.safetensors:  68%|██████▊   | 371/545 [00:02<00:01, 129.33it/s]
Loading anima_baseV10.safetensors:  71%|███████   | 386/545 [00:02<00:01, 129.69it/s]
Loading anima_baseV10.safetensors:  74%|███████▎  | 401/545 [00:02<00:01, 132.73it/s]
Loading anima_baseV10.safetensors:  76%|███████▋  | 416/545 [00:02<00:01, 125.36it/s]
Loading anima_baseV10.safetensors:  93%|█████████▎| 509/545 [00:03<00:00, 331.04it/s]
                                                                                     
2026-06-24 03:36:02 INFO     Loaded DiT model from                weights.py:203
                             E:/AI/ComfyUI-aki-v3/ComfyUI/models/               
                             diffusion_models/anima_baseV10.safet               
                             ensors, unexpected missing keys: 0,                
                             unexpected keys: 0                                 
                    INFO     static_token_count=4096                train.py:567
import network module: networks.lora_anima
2026-06-24 03:36:03 INFO     create LoRA network. base dim        network.py:503
                             (rank): 16, alpha: 16.0                            
                    INFO     neuron dropout: p=None, rank         network.py:506
                             dropout: p=None, module dropout:                   
                             p=None                                             
                    INFO     create LoRA for Text Encoder 1:      network.py:882

Creating TE1 LoRA:   0%|          | 0/196 [00:00<?, ?it/s]
                                                          
                    INFO     create LoRA for Text Encoder 1: 196  network.py:889
                             modules.                                           

Creating DiT LoRA:   0%|          | 0/196 [00:00<?, ?it/s]
Creating DiT LoRA:  73%|███████▎  | 144/196 [00:00<00:00, 1434.85it/s]
                                                                      
                    INFO     create LoRA for Anima DiT: 196       network.py:903
                             modules.                                           
                    INFO     use_custom_down_autograd: enabled on factory.py:135
                             392 LoRA-family modules (saves                     
                             ~32-128 MiB/Linear of fp32                         
                             activation per step)                               
                    INFO     enable LoRA for DiT: 196 modules    network.py:2685
prepare optimizer, data loader etc.
                    INFO     use Prodigy optimizer |           optimizers.py:220
                             {'decouple': True,                                 
                             'weight_decay': 0.01,                              
                             'use_bias_correction': True,                       
                             'd_coef': 2.0}                                     
Using decoupled weight decay
override steps. steps for 30 epochs is
Anima: compiled _run_blocks with backend=inductor, mode=None
2026-06-24 03:36:04 INFO     compile_core: _run_blocks compiled    train.py:1892
                             (backend=inductor, mode=None)                      
running training
  num train images * repeats
  num validation images * repeats
  num reg images
  num batches per epoch
  num epochs
  batch size per device
  gradient accumulation steps
  total optimization steps
2026-06-24 03:36:08 INFO     unet dtype: torch.bfloat16, device:     loop.py:213
                             cuda:0                                             
                    INFO     text_encoder [0] dtype: torch.bfloat16, loop.py:219
                             device: cuda:0                                     

steps:   0%|          | 0/150 [00:00<?, ?it/s]
epoch 1/30

2026-06-24 03:36:24 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 1                                           
2026-06-24 03:36:24 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 1                                           
2026-06-24 03:36:24 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 1                                           
2026-06-24 03:36:24 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 1                                           

steps:   1%|          | 1/150 [01:43<4:15:47, 103.00s/it, avr_loss=0.115]
steps:   1%|▏         | 2/150 [01:48<2:13:32, 54.14s/it, avr_loss=0.107] 
steps:   2%|▏         | 3/150 [01:53<1:32:53, 37.91s/it, avr_loss=0.0982]
steps:   3%|▎         | 4/150 [01:59<1:12:28, 29.78s/it, avr_loss=0.0968]
steps:   3%|▎         | 5/150 [02:03<59:50, 24.76s/it, avr_loss=0.0942]  
saving checkpoint: D:/AI/anima_lora_train/lora-scripts-next/output\NaoJMaru-000001.safetensors

epoch 2/30

2026-06-24 03:38:29 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 2                                           
2026-06-24 03:38:29 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 2                                           
2026-06-24 03:38:29 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 2                                           
2026-06-24 03:38:29 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 2                                           

steps:   4%|▍         | 6/150 [02:27<58:48, 24.50s/it, avr_loss=0.0868]
steps:   5%|▍         | 7/150 [02:32<51:55, 21.79s/it, avr_loss=0.0845]
steps:   5%|▌         | 8/150 [02:37<46:35, 19.69s/it, avr_loss=0.0841]
steps:   6%|▌         | 9/150 [02:42<42:28, 18.07s/it, avr_loss=0.0924]
steps:   7%|▋         | 10/150 [02:46<38:56, 16.69s/it, avr_loss=0.0907]
saving checkpoint: D:/AI/anima_lora_train/lora-scripts-next/output\NaoJMaru-000002.safetensors

epoch 3/30

2026-06-24 03:39:11 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 3                                           
2026-06-24 03:39:11 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 3                                           
2026-06-24 03:39:11 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 3                                           
2026-06-24 03:39:11 INFO     epoch is incremented. current_epoch: 0, base.py:229
                             epoch: 3                                           

steps:   7%|▋         | 11/150 [03:08<39:37, 17.10s/it, avr_loss=0.0925]
steps:   8%|▊         | 12/150 [03:13<37:00, 16.09s/it, avr_loss=0.095] 
steps:   9%|▊         | 13/150 [03:18<34:47, 15.24s/it, avr_loss=0.0953]
steps:   9%|▉         | 14/150 [03:22<32:51, 14.50s/it, avr_loss=0.0872]
steps:  10%|█         | 15/150 [03:27<31:07, 13.83s/it, avr_loss=0.0964]
saving checkpoint: D:/AI/anima_lora_train/lora-scripts-next/output\NaoJMaru-000003.safetensors

**Comment by wochenlong** [comment]
根因调查与多次修复尝试已整理到 #162（前端参数预览不显示 \enable_preview\、schemastery intersect+union 序列化、dist 补丁与后端兜底现状）。本 Issue 保持 open，训练侧预览图问题与 #162 同源，待协作者一起闭环。

**Comment by soutsukimaru** [comment]
预览图已确认修复，谢谢！
